web-app-pix2info-python/src/backend/docai.py

```diff
@@ -0,0 +1,278 @@
+    Processor,
+    ProcessRequest,
+    RawDocument,
+)
+from google.protobuf.json_format import MessageToJson  # type: ignore
+
+from .processors import (
+    DEMO_PROCESSING_LOCATIONS,
+    DEMO_PROCESSOR_TYPES,
+    SAMPLE_PROCESSING_LOCATION,
+    decode_processor_info,
+    encode_processor_info,
+)
+from .render import tiff_container_for_images, will_render_entity
+
+DocumentStream: TypeAlias = BytesIO
+MimeType: TypeAlias = str
+DocumentIO: TypeAlias = tuple[DocumentStream, MimeType]
+DocumentJson: TypeAlias = str
+DocumentData: TypeAlias = tuple[Document, DocumentJson]
+
+
+def get_client(location: str) -> DocumentProcessorServiceClient:
+    """Return a Document AI client."""
+    client_options = dict(api_endpoint=f"{location}-documentai.googleapis.com")
+    return DocumentProcessorServiceClient(client_options=client_options)
+
+
+def process_document(
+    file: BinaryIO,
+    mime_type: str,
+    project: str,
+    location: str,
+    processor_id: str,
+) -> Document:
+    """Analyze the input file with Document AI."""
+    client = get_client(location)
+    raw_document = RawDocument(content=file.read(), mime_type=mime_type)
+    name = client.processor_path(project, location, processor_id)
+
+    request = ProcessRequest(
+        raw_document=raw_document,
+        name=name,
+        skip_human_review=True,
+    )
+    response = client.process_document(request)
+
+    return response.document
+
+
+def process_documents(
+    documents: Sequence[DocumentIO],
+    project: str,
+    location: str,
+    processor_id: str,
+) -> DocumentData:
+    """Analyze the input file(s) with Document AI.
+
+    Return both the Document AI object and its json serialization (which can be viewed/cached).
+    """
+    if not documents:
+        raise ValueError("At least one document is expected")
+
+    if len(documents) == 1:
+        image_io, mime_type = documents[0]
+    else:
+        images = [file for (file, _) in documents]
+        image_io, mime_type = tiff_container_for_images(images)
+
+    document = process_document(image_io, mime_type, project, location, processor_id)
+    json = MessageToJson(
+        Document.pb(document),
+        preserving_proto_field_name=True,  # Same as Python (switch if camel case is preferred)
+        ensure_ascii=False,  # Don't escape characters (easier to read + smaller files)
+    )
+
+    return document, json
+
+
+def process_sample(
+    sample_name: str,
+    samples_root: Path,
+    samples_paths: Sequence[str],
+    project: str,
+    processor_name: str,
+) -> DocumentData:
+    """Analyze the sample file(s) with Document AI.
+
+    Analyses are cached. This allows deploying a demo with cached samples only.
+    """
+    json_path = samples_root.joinpath(processor_name, f"{sample_name}.json")
+    if json_path.exists():
+        # Use cached json serialization
+        json = json_path.read_text(encoding="utf-8")
+        document = cast(Document, Document.from_json(json))
```

Comment:
Not sure if you need to cast this. `Document.from_json()` returns a `Document` type

Reply:
It returns a document instance but it's typed with the base class `proto.Message`. The cast is to silence `mypy`.